content/cif-connector/src/main/content/jcr_root/libs/commerce/gui/components/authoring/editor/pagepreview/clientlibs/pagepreview.js

```diff
@@ -0,0 +1,81 @@
+/*******************************************************************************
+ *
+ *    Copyright 2019 Adobe. All rights reserved.
+ *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License. You may obtain a copy
+ *    of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software distributed under
+ *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ *    OF ANY KIND, either express or implied. See the License for the specific language
+ *    governing permissions and limitations under the License.
+ *
+ ******************************************************************************/
+
+/*
+ * Actions for the preview of product detail pages.
+ */
+
+(function(window, document, Granite, $) {
+    'use strict';
+
+    var relPdpPreview = '.cq-commerce-pdp-preview-activator';
+
+    var handlePdpPreview = function(e, data) {
+        if (!data) {
+            return;
+        }
+
+        var selections = data.selections;
+        var productSlug = null;
+        if (Array.isArray(selections)) {
+            if (selections.length > 0 && selections[0]) {
+                productSlug = selections[0].value;
+            }
+        } else if (selections) {
+            productSlug = selections.value;
+        }
+
+        var url = Granite.author.ContentFrame.location;
+        var previewUrl = createPreviewUrl(url, productSlug);
+
+        if (previewUrl) {
+            window.open(previewUrl);
+        }
+    };
+
+    var createPreviewUrl = function(url, slug) {
+        if (!url || !slug) {
+            return null;
+        }
+
+        var indLastSlash = url.lastIndexOf('/');
+        if (indLastSlash < 0 || indLastSlash === url.length - 1) {
+            return null;
+        }
+
+        var urlPath = url.slice(0, indLastSlash + 1);
+        var urlName = url.substring(indLastSlash + 1);
+        var indHtml = urlName.lastIndexOf('.html');
+        var previewUrl = null;
+        if (indHtml > -1) {
+            var indSelector = urlName.slice(0, indHtml).lastIndexOf('.');
+            if (indSelector > -1) {
+                previewUrl = urlPath + urlName.slice(0, indSelector) + '.' + slug + '.html';
+            } else {
+                previewUrl = urlPath + urlName.slice(0, indHtml) + '.' + slug + '.html';
+            }
+        }
+
+        return previewUrl;
+    };
+
+    Granite.$(document).on('cifProductPickerSelection', relPdpPreview, handlePdpPreview);
+
+    if (window.CifTesting) {
```

Comment:
If this is used only for testing it shouldn't be here. I understand that you did this in order to be able to inject a mock Granite instance. Another way to do this is this:
- declare this whole function in its own scope:

```javascript
window.CIF.PagePreview = function(window, document, Granite) { ... }
```

- immediately call it with 
```javascript
window.CIF.PagePreview(window,document,Granite)
```

...and in your tests you can call it with your own mock parameters.

Reply:
@dplaton , I considered that and favored the solution above in order to avoid creating a new scope which is what Granite UI seems to prefer.
I would say it's a small price to pay for being consistent with Granite.

Reply:
I agree with Daniel here. Mixing testing and production code is a bad practise and should be avoided at any cost.

Reply:
+1

Reply:
>I would say it's a small price 
I don't consider it small. You wouldn't mix testing code in your Java classes either. The most "mixing" I've seen people doing (incl myself) is using `protected` methods instead of `private`.

>[...] which is what Granite UI seems to prefer.
Granite UI has almost zero unit testing (and I say almost because there might be some that I don't know of) so they didn't need to write unit-testable code. However, we do aim to have unit-tests so we must write our code in such a manner. The bare minimum we can do is put our functions in the global namespace (`window.CIF` is used by the non-React client-side components as well).